crates/net/network/src/transactions.rs

```diff
@@ -524,47 +545,204 @@ where
 
-            if num_already_seen > 0 {
-                self.metrics.messages_with_already_seen_hashes.increment(1);
-                trace!(target: "net::tx", num_hashes=%num_already_seen, ?peer_id, client=?peer.client_version, "Peer sent already seen hashes");
-            }
+        // only send request for hashes to idle peer, otherwise buffer hashes storing peer as
+        // fallback
+        if !self.transaction_fetcher.is_idle(peer_id) {
+            trace!(target: "net::tx",
+                peer_id=abbrev_hex(peer_id.as_ref()),
+                hashes=abbrev_hex_hash_list(&hashes),
+                "buffering hashes announced by busy peer"
+            );
+
+            self.transaction_fetcher.buffer_hashes(hashes, Some(peer_id));
+            return
+        }
+        // demand recommended soft limit on response, however the peer may enforce an arbitrary
+        // limit on the response (2MB)
+        if let Some(surplus_hashes) =
+            self.transaction_fetcher.pack_hashes(&mut hashes, peer_id, is_valid_for_version_eth68)
+        {
+            trace!(target: "net::tx",
+                peer_id=abbrev_hex(peer_id.as_ref()),
+                surplus_hashes=abbrev_hex_hash_list(&surplus_hashes),
+                "some hashes in announcement from peer didn't fit in `GetPooledTransactions` request, buffering surplus hashes"
+            );
+            self.transaction_fetcher.buffer_hashes(surplus_hashes, Some(peer_id));
+        }
+
+        trace!(target: "net::tx",
+            peer_id=abbrev_hex(peer_id.as_ref()),
+            hashes=abbrev_hex_hash_list(&hashes),
+            "sending hashes in `GetPooledTransactions` request to peer's session"
+        );
+
+        // request the missing transactions
+        let metrics = &self.metrics;
+        if let Some(failed_to_request_hashes) =
+            self.transaction_fetcher.request_transactions_from_peer(hashes, peer, || {
+                metrics.egress_peer_channel_full.increment(1)
+            })
+        {
+            debug!(target: "net::tx",
+                peer_id=abbrev_hex(peer_id.as_ref()),
+                hashes=%abbrev_hex_hash_list(&failed_to_request_hashes),
+                "sending `GetPooledTransactions` request to peer's session failed, buffering hashes"
+            );
+            self.transaction_fetcher.buffer_hashes(failed_to_request_hashes, Some(peer_id));
+            return
         }
 
         if num_already_seen > 0 {
+            self.metrics.messages_with_already_seen_hashes.increment(1);
+            trace!(target: "net::tx", num_hashes=%num_already_seen, ?peer_id, client=?peer.client_version, "Peer sent already seen hashes");
             self.report_already_seen(peer_id);
         }
     }
 
+    // Requests buffered hashes for which an idle peer exists.
+    fn request_buffered_hashes(&mut self) {
+        loop {
+            let mut hashes = vec![];
+            let Some(peer_id) = self.pop_any_idle_peer(&mut hashes) else {
```

Comment:
I'd appreciate a few more docs for these two functions.

If I read that correctly then this tries to issue many fallback requests as possible to matching peers but only requesting 1 hash at a time?

Reply:
it tries to fill the request for that peer, so more than one hash at a time, if the peer is listed as fallback for more than one peer.